File: src/test/java/com/github/sszuev/graphs/CommonRunners.java
package com.github.sszuev.graphs;

import org.apache.jena.graph.Graph;
import org.junit.jupiter.api.Assertions;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

public class CommonRunners {
    public static final long EXECUTE_TIMEOUT_MS = 60 * 1000L;

    public static void runTestScenario(
            Graph graph,
            int numberOfThreads,
            int innerIterations,
            GraphScenario scenario) {
        runTestScenario(graph, numberOfThreads, innerIterations, (g, tid, iid) -> scenario.run(g));
    }

    public static void runTestScenario(
            Graph graph,
            int numberOfThreads,
            int innerIterations,
            DetailedGraphScenario scenario) {
        ExecutorService executor = Executors.newFixedThreadPool(numberOfThreads);
        CyclicBarrier barrier = new CyclicBarrier(numberOfThreads);
        List<Future<Void>> tasks = new ArrayList<>();
        for (int ti = 1; ti <= numberOfThreads; ti++) {
            final int tid = ti;
            tasks.add(executor.submit(() -> {
                for (int ii = 0; ii < innerIterations; ii++) {
                    try {
                        barrier.await(EXECUTE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
                        scenario.run(graph, tid, ii);
                    } catch (Exception ex) {
                        executor.shutdownNow();
                        throw ex;
                    }
                }
                return null;
            }));
        }
        executor.shutdown();
        try {
            Assertions.assertTrue(executor.awaitTermination(EXECUTE_TIMEOUT_MS, TimeUnit.MILLISECONDS));
        } catch (InterruptedException e) {
            throw new AssertionError(e);
        }
        AssertionError error = new AssertionError();
        tasks.forEach(task -> {
            try {
                task.get(EXECUTE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
            } catch (Exception ex) {
                error.addSuppressed(ex);
            }
        });
        if (error.getSuppressed().length > 0) {
            throw error;
        }
    }

    public interface GraphScenario {
        void run(Graph graph);
    }

    public interface DetailedGraphScenario {
        void run(Graph graph, int tid, int iid) throws Exception;
    }
}

File: src/test/kotlin/testutils/CommonRunners.kt
package com.github.sszuev.graphs.testutils

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import org.apache.jena.graph.Graph
import java.util.concurrent.Callable
import java.util.concurrent.CyclicBarrier
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

internal const val EXECUTE_TIMEOUT_MS = 60 * 1000L

internal fun runTestScenario(
    graph: Graph,
    numberOfThreads: Int,
    innerIterations: Int,
    scenario: (Graph) -> Unit
) {
    runTestScenario(graph, numberOfThreads, innerIterations) { g, _, _ ->
        scenario(g)
    }
}

internal fun runTestScenario(
    graph: Graph,
    numberOfThreads: Int,
    innerIterations: Int,
    scenario: (Graph, tid: Int, iid: Int) -> Unit
) {
    val executor = Executors.newFixedThreadPool(numberOfThreads)
    val barrier = CyclicBarrier(numberOfThreads)
    val tasks = (1..numberOfThreads).map { ti ->
        executor.submit {
            repeat(innerIterations) { ii ->
                barrier.await(EXECUTE_TIMEOUT_MS, TimeUnit.MILLISECONDS)
                try {
                    scenario(graph, ti, ii)
                } catch (ex: Exception) {
                    executor.shutdownNow()
                    throw ex
                }
            }
        }
    }
    executor.shutdown()
    executor.awaitTermination(EXECUTE_TIMEOUT_MS, TimeUnit.MILLISECONDS)
    val error = AssertionError()
    tasks.forEach {
        try {
            it.get(EXECUTE_TIMEOUT_MS, TimeUnit.MILLISECONDS)
        } catch (ex: Exception) {
            error.addSuppressed(ex)
        }
    }
    if (error.suppressed.isNotEmpty()) {
        throw error
    }
}

internal fun ExecutorService.runAll(tasks: Collection<Callable<Unit>>) {
    val futures = invokeAll(tasks)
    shutdown()
    awaitTermination(EXECUTE_TIMEOUT_MS, TimeUnit.MILLISECONDS)
    val error = AssertionError("ExecutorService.runAll::fail")
    futures.forEach {
        try {
            it.get()
        } catch (e: Exception) {
            shutdownNow()
            error.addSuppressed(e)
        }
    }
    if (error.suppressed.isNotEmpty()) {
        if (error.suppressed.size == 1) {
            throw error.suppressed[0]
        } else {
            throw error
        }
    }
}

internal fun CoroutineDispatcher.runAll(tasks: Collection<Callable<Unit>>) {
    runBlocking(this) {
        withTimeout(EXECUTE_TIMEOUT_MS) {
            val jobs = tasks.map {
                launch { it.call() }
            }
            jobs.joinAll()
        }
    }
}